### PR TITLE
issue #1352 - introduce fhir-smart component

### PR DESCRIPTION
**Pull request by lmsurpre**
in this initial version, fhir-smart includes a single PersistenceInterceptor which uses java-jwt to decode
a passed token, parses the patient context from a "patient_id" claim, parses the granted scope string,
and checks these values against the interaction being performed.

Signed-off-by: Lee Surprenant <lmsurpre@us.ibm.com>

### CODE REVIEW

File: fhir-smart/pom.xml [hunk 1]
```diff
@@ -0,0 +1,77 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- THIS PRODUCT CONTAINS RESTRICTED MATERIALS OF IBM 5724-H88, 5724-J08,
+    5724-I63, 5655-W65, COPYRIGHT International Business Machines Corp., 2014
+    All Rights Reserved * Licensed Materials - Property of IBM US Government
+    Users Restricted Rights - Use, duplication or disclosure restricted by GSA
+    ADP Schedule Contract with IBM Corp. -->
```
Comment (prb112) [inline fhir-smart/pom.xml:6]: Copyright?

File: fhir-smart/pom.xml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>com.ibm.fhir</groupId>
+        <artifactId>fhir-parent</artifactId>
+        <version>4.3.2-SNAPSHOT</version>
```
Comment (prb112) [inline fhir-smart/pom.xml:16]:
```suggestion
        <version>4.4.0-SNAPSHOT</version>
```

File: fhir-smart/src/test/java/com/ibm/fhir/smart/test/AuthzPolicyBackupEnforcementTest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
+package com.ibm.fhir.smart.test;
```
Comment (prb112) [inline fhir-smart/src/test/java/com/ibm/fhir/smart/test/AuthzPolicyBackupEnforcementTest.java:1]: copyright

File: fhir-smart/src/test/java/com/ibm/fhir/smart/test/AuthzPolicyBackupEnforcementTest.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        interceptor = new AuthzPolicyEnforcementPersistenceInterceptor();
+
+        patient = TestUtil.readExampleResource("json/ibm/minimal/Patient-1.json");
+        patient = patient.toBuilder()
+                .id("11111111-1111-1111-1111-111111111111")
+                .build();
+    }
+
+    @Test
+    public void testUserInfoHeader() throws Exception {
```
Comment (prb112) [inline fhir-smart/src/test/java/com/ibm/fhir/smart/test/AuthzPolicyBackupEnforcementTest.java:33]: Negative test too?

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        Objects.requireNonNull(contextIds, "contextIds");
+
+        String resourceType = resource.getClass().getSimpleName();
+        Map<ContextType, List<Scope>> approvedScopeMap = approvedScopes.stream()
+                // First filter the list to only scopes which grant the required permissions on the passed resourceType
+                .filter(s -> s.getResourceType() == ResourceType.ValueSet.RESOURCE ||
+                        s.getResourceType().value().equals(resourceType))
+                .filter(s -> hasPermission(s.getPermission(),requiredPermission))
+                // Then group the scopes by their context type
+                .collect(Collectors.groupingBy(s -> s.getContextType()));
```
Comment (prb112) [inline fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:128]: Given John's issues with streams, how does this perform?

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+                                }
+                                return;
+                            }
+                        }
+                    } catch (Exception e) {
+                        log.log(Level.WARNING, "Unexpected exception while processing inclusionCriteria '" + searchParm +
+                                "' in the Patient compartment for resource type " + resourceType);
+                    }
+                }
+            } catch (FHIRSearchException e) {
```
Comment (prb112) [inline fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:167]: FHIRSearchException... shouldn't this be an operation exception?

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+            return true;
+        } else {
+            return grantedPermission == requiredPermission;
+        }
+    }
+
+    private String getAccessToken() throws FHIRPersistenceInterceptorException {
+        List<String> list = FHIRRequestContext.get().getHttpHeaders().get("Authorization");
+        if (list.size() != 1) {
+            throw new FHIRPersistenceInterceptorException("Request must contain exactly one Authorization header.");
```
Comment (prb112) [inline fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:221]: interesting check...

File: fhir-smart/.gitignore [hunk 8]
```diff
@@ -0,0 +1,6 @@
+/.apt_generated/
+/.settings/
+/target/
+/test-output/
+.factorypath
+*.log
```
Comment (prb112) [inline fhir-smart/.gitignore:6]: should this perhaps go in a not top-level folder? 
or should this be fhir-auth-smart? 
Is it just going to be scoped to auth intercepting

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+            scopeStrings = Arrays.asList(claim.asString().split("\\s+"));
+        } else {
+            log.fine("Found scope claim was expected to be a string but is not; processing as a list");
+            scopeStrings = claim.asList(String.class);
+        }
+
+        return scopeStrings.stream()
+                .filter(s -> s.matches(Scope.SCOPE_STRING_REGEX))
+                .map(s -> new Scope(s))
+                .collect(Collectors.toList());
```
Comment (prb112) [inline fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:251]: small set may be best to have a for loop.... same performance question

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+        }
+
+        String patientId = claim.asString();
+        if (patientId == null) {
+            log.fine("Found patient_id claim was expected to be a string but is not; processing as a list");
+            return claim.asList(String.class);
+        }
+
+        return Collections.singletonList(patientId);
+    }
```
Comment (prb112) [inline fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:267]: Are you anticipating more than one id?

File: fhir-smart/pom.xml (hunk 1 above)
Comment (lmsurpre) [reply fhir-smart/pom.xml]: fixed

File: fhir-smart/src/test/java/com/ibm/fhir/smart/test/AuthzPolicyBackupEnforcementTest.java (hunk 3 above)
Comment (lmsurpre) [reply fhir-smart/src/test/java/com/ibm/fhir/smart/test/AuthzPolicyBackupEnforcementTest.java:1]: fixed

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java (hunk 5 above)
Comment (lmsurpre) [reply fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:128]: there will typically only be one or two scope strings in the JWT, so I'm really not worried about it

File: fhir-smart/src/test/java/com/ibm/fhir/smart/test/AuthzPolicyBackupEnforcementTest.java (hunk 4 above)
Comment (lmsurpre) [reply fhir-smart/src/test/java/com/ibm/fhir/smart/test/AuthzPolicyBackupEnforcementTest.java:33]: yes, needs much more comprehensive tests in general I think

File: fhir-smart/pom.xml (hunk 2 above)
Comment (lmsurpre) [reply fhir-smart/pom.xml:11]: Sure, I can accept your other PR, rebase, and then make this change.

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java (hunk 9 above)
Comment (lmsurpre) [reply fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:251]: same response.  can change if needed, but I think its fine.

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java (hunk 10 above)
Comment (lmsurpre) [reply fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:267]: Wanted to leave the door open for it, but could refactor to assume a single patientId is in-scope if desired.

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java (hunk 6 above)
Comment (lmsurpre) [reply fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:167]: this is whats thrown from `CompartmentUtil.getCompartmentResourceTypeInclusionCriteria`.  Potentially that could get refactored to throw something else, but I think its beyond the scope of this PR.  Please re-open if you disagree.

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java (hunk 7 above)
Comment (lmsurpre) [reply fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:221]: just trying to be defensive there, not attached to it.  re-open if you think it should be changed.

File: fhir-smart/.gitignore (hunk 8 above)
Comment (lmsurpre) [reply fhir-smart/.gitignore:6]: we already have a top-level `.ignore` that covers what we'll need...I'll empty this one out

Comment (lmsurpre) [reply fhir-smart/.gitignore]: done

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java (hunk 10 above)
Comment (prb112) [reply fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:267]: No refactoring. This was a question given the multiple potential ids.

File: fhir-smart/src/test/java/com/ibm/fhir/smart/test/AuthzPolicyBackupEnforcementTest.java (hunk 4 above)
Comment (prb112) [reply fhir-smart/src/test/java/com/ibm/fhir/smart/test/AuthzPolicyBackupEnforcementTest.java]: No problem - is this coming in a separate issue? Should we track an issue for it?

File: fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java (hunk 6 above)
Comment (prb112) [reply fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java:167]: this was more of a confirmation question.

File: fhir-smart/src/test/java/com/ibm/fhir/smart/test/AuthzPolicyBackupEnforcementTest.java (hunk 4 above)
Comment (lmsurpre) [reply fhir-smart/src/test/java/com/ibm/fhir/smart/test/AuthzPolicyBackupEnforcementTest.java]: I just posted another commit here...can you please have a look?

**Pull request merged**